src/system/SystemSocketWatcherSelect.h

```diff
@@ -0,0 +1,79 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implements a SystemSocketWatcher class for select()-based platforms.
+ *      This is not a full SystemSocketWatcher implementation, as there is
+ *      still select()-specific code in platform
+ */
+
+#pragma once
+
+#include <sys/select.h>
+
+#include <support/BitFlags.h>
+#include <system/SystemSockets.h>
+
+namespace chip {
+namespace System {
+
+class Layer;
+
+class SystemSocketWatchState
+{
+public:
+    void Init() {}
+    void Shutdown() {}
+};
+
+/*
+ * Interface to request action on I/O availability.
+ *
+ * EndPoint subclasses call these functions to specify the conditions they are currently interested in.
+ *
+ * For the CHIP_SYSTEM_CONFIG_SOCKETS_USE_SELECT==1 case, these functions (implemented inline below) will
+ * set mRequestIO for inspection by the platform layer, which uses InetLayer::HandleSelectResult() to
+ * call HandlePendingIO() in the requested state, after setting mRequestIO to indicate the conditions
+ * that are satisfied.
+ *
+ * For the CHIP_SYSTEM_CONFIG_SOCKETS_USE_SELECT==0 case, the Watch implementation and the platform layer
+ * are required to call the defined WatchCallback, with mPendingIO indicating the watch condition(s)
+ * that are satisfied.
+ */
+class SystemSocketWatcher : public System::SystemSocketWatcherBasis
+{
+public:
+    void Init(Layer & systemLayer) { mRequestIO.ClearAll(); }
+    void Open(int fd) {}
+    void Close() { mRequestIO.ClearAll(); }
+    void SetReadable() { mRequestIO.Set(System::SocketEventFlag::kRead); }
+    void SetWritable() { mRequestIO.Set(System::SocketEventFlag::kWrite); }
+    void ClearReadable() { mRequestIO.Clear(System::SocketEventFlag::kRead); }
+    void ClearWritable() { mRequestIO.Clear(System::SocketEventFlag::kWrite); }
+
+    // Platform is currently hard-coded to use InetLayer::HandleSelectResult().
```

Comment:
Will we able to remove this discrepancy and make both work with SetCallback & remove HandleSelectResult?

Reply:
Yes, I think that should be easy; I just put it off. The remaining #if-select() code would be the `PrepareSelect`/`HandleSelectResult` stuff in //system, which looks a bit harder to refactor, but probably possible — if anybody's still going to want to use select() in the long run.